nemo/collections/vlm/qwen2vl/model/base.py

```diff
@@ -0,0 +1,1040 @@
+        language_embeddings = None
+
+        language_seq_len = input_ids.shape[1]
+        # chunk if input seq_len > _language_max_sequence_length
+        if language_seq_len > self._language_max_sequence_length:
+            input_ids = input_ids[:, : self._language_max_sequence_length]
+            if position_ids is not None:
+                position_ids = position_ids[:, :, : self._language_max_sequence_length]
+
+            if labels is not None and labels.shape[1] > self._language_max_sequence_length:
+                labels = labels[:, : self._language_max_sequence_length]
+                loss_mask = loss_mask[:, : self._language_max_sequence_length]
+
+        # Pipeline parallel expects fixed input size. Check if we need to pad.
+        if self._language_is_pipeline_parallel and language_seq_len < self._language_max_sequence_length:
+            padded_seq_len = self._language_max_sequence_length - language_seq_len
+            input_ids = torch.nn.functional.pad(input_ids, (0, padded_seq_len))
+            if position_ids is not None:
+                position_ids = torch.nn.functional.pad(position_ids, (0, padded_seq_len))
+
+        if position_ids is None and input_ids is not None:
+            position_ids, _ = self.get_rope_index(input_ids, image_grid_thw, video_grid_thw, attention_mask)
+
+        # Create the language_embeddings (if this is the first language model stage).
+        if self.pre_process:
+
+            # Note: This adds absolute position embedding but not RoPE.
+            # Each image is counted as one position.
+            # RoPE is added in language_model forward. Each image embedding is one position.
+            if self.sequence_parallel_lm:
+                # Pad to nearest multiple of TP world size for embedding.
+                tp_world_size = ps.get_tensor_model_parallel_world_size()
+                padded_seq_len = (
+                    int((input_ids.shape[1] + tp_world_size - 1) // tp_world_size * tp_world_size) - input_ids.shape[1]
+                )
+                if padded_seq_len != 0:
+                    input_ids = torch.nn.functional.pad(input_ids, (0, padded_seq_len))
+                    if position_ids is not None:
+                        position_ids = torch.nn.functional.pad(position_ids, (0, padded_seq_len))
+
+            input_ids_text = input_ids.clone()
+            # MultiModal Token indices are assumed to be values
+            input_ids_text[input_ids_text < 0] = 0
+
+            language_embeddings = self.language_model.embedding(
+                input_ids=input_ids_text, position_ids=None
+            )  # [decoder_seq_len, b, h_language]
+
+            if self.sequence_parallel_lm:
+                # Gather the language embeddings back.
+                # We use the full embedding to insert image embeddings
+                # and then scatter to avoid load imbalance.
+                language_embeddings = gather_from_sequence_parallel_region(
```

Comment:
this is no longer needed. Just add https://github.com/NVIDIA/NeMo/blob/c22b3da71e108f7d1b50cb79abcacdfaa4c48aa6/nemo/collections/vlm/neva/model/base.py#L282